accepted/future-releases/0546-patterns/exhaustiveness.md

```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
 
-To calculate if the intersection `I` of spaces `L` and `R` is empty:
+*For example, the expansion of `FutureOr<bool?>` is:*
 
-1.  If `L` or `R` is empty, then `I` is empty.
+```
+Future<bool?>
+true
+false
+null
+```
 
-2.  If `L` or `R` is a union, then `I` is empty if the intersection of every
-    arm of the union with the other operand is empty.
+### Overapproximation
 
-3.  Otherwise, we're intersecting two object spaces.
+The *overapproximation* of a type `T` is `T` with all type variables replaced
+with their *defaults*:
 
-    1.  If neither object's type is a subtype of the other then `I` is empty.
-        They are unrelated types with no (known) intersection.
+1.  If `T` is in a contravariant position, then the default is `Never`.
 
-    2.  Otherwise, go through the fields. If the intersection of any
-        corresponding pair of fields is empty, then `I` is empty.
+2.  Otherwise, the default is the bound if there is one.
 
-    3.  Otherwise, `I` is not empty. We found two objects where one is a
-        subtype of the other and the fields (if any) have at least some overlap.
+3.  Otherwise, the default is `Object?`.
 
-## Conclusion
+The overapproximation of a type reliably contains all values that any possible
+instantiation of the type could contain. This lets us calculate exhaustiveness
+soundly even in the context of parameter types whose concrete instantiations
+aren't known.
 
-The basic summary is:
+### Filtering by a space
 
-1.  We take the language's notion of static types and patterns and lift them to
-    a simpler but more expressive model of spaces.
+We're given a single `valueSpace`, a `value` worklist for further value
+constraints to explore, and a set of `cases` worklists that may match `space`.
+Next, we discard any cases that won't be helpful for exhaustiveness given what
+we know now. If it's *possible* for any `value` in value to *not* match the
+first space in a given worklist in cases, then that case can't guarantee
+exhaustiveness, so we remove it.
 
-2.  We define space subtraction and intersection.
+1.  Let `remaining` be an empty list of worklists.
 
-3.  Using that, we can define exhaustiveness and reachability as a simple series
-    of subtractions over the spaces for a switch's value type and case patterns.
+2.  Let `caseFirstSpaces` be an empty list of spaces.
 
-The [prototype][] has a number of tests that try to cover all of the interesting
-cases, though there are so many ways that object spaces can vary and compose
-that it's hard to be sure everything is tested.
+3.  For each `worklist` in `cases`:
 
-## Next steps
+    1.  Dequeue the first case union in `worklist` to `caseUnion`.
 
-I haven't proven that the algorithm is correct. I also haven't proven any bounds
-on its performance. A fairly large test covering every combination of four
-fields seems to be snappy (and certainly was not with earlier versions of this
-algorithm).
+    1.  For each `case` space in `caseUnion`:
```

Comment:
The item number is off here and in the sublist below (not that it matters to the markdown renderer).